File: web/astro/transits.js
// Транзиты: как градус в небе задевает натальную карту.
//
// Транзит — это встреча, а не свойство карты: планета в небе подходит к
// градусу, который в карте чем-то занят. Поэтому считается не «что
// происходит», а «за что именно в этой карте цепляется вот этот градус».

import { MAJOR, ASPECTS } from './aspects.js';
import { houseOf } from './houses.js';
import { rulerOf, TRADITIONAL } from './rulers.js';
import { norm180, signIndex } from './zodiac.js';

// Орбисы транзитов уже натальных: планета проходит градус за дни или
// недели, и широкий орбис размазал бы событие на месяцы.
export const TRANSIT_ORBS = {
  conjunction: 3, opposition: 3, square: 3, trine: 3, sextile: 2,
};

export const CUSP_ORB = 2;

// Точным считается контакт в пределах четверти градуса — это примерно
// сутки хода для Луны и недели для внешних планет.
const EXACT = 0.25;

export function ruledHouses(chart) {
  // Дом управляется телом, которому принадлежит знак на его куспиде.
  // Это и переводит «задета планета» в «задета тема»: транзит к
  // управителю второго дома говорит о деньгах, даже если сама планета
  // стоит в седьмом. Управители берутся из той же схемы, по которой
  // построена карта.
  const cusps = chart.houses.get(chart.houseSystem).cusps;
  const mapping = new Map();
  for (let house = 1; house <= 12; house += 1) {
    const ruler = rulerOf(signIndex(cusps[house - 1]), chart.rulerScheme || TRADITIONAL);
    if (!mapping.has(ruler)) mapping.set(ruler, []);
    mapping.get(ruler).push(house);
  }
  return mapping;
}

export function examine(chart, longitude, transitName = 'транзит', options = {}) {
  const {
    orbs = TRANSIT_ORBS,
    cuspOrb = CUSP_ORB,
    aspects = MAJOR,
  } = options;

  const hasHouses = chart.exactTime !== false;
  const hits = [];

  const check = (natal, natalLongitude, kind) => {
    for (const aspect of aspects) {
      const limit = orbs[aspect.key];
      if (!limit) continue;
      const delta = norm180(longitude - natalLongitude);
      const target = delta >= 0 ? aspect.angle : -aspect.angle;
      const orb = Math.abs(norm180(delta - target));
      if (orb <= limit) {
        hits.push({
          transit: transitName, natal, natalKind: kind, aspect, orb,
          exact: orb <= EXACT, natalLongitude,
          strength: Math.max(0, 1 - orb / limit),
        });
        return; // ближайший аспект для пары найден
      }
    }
  };

  for (const [key, position] of chart.positions) {
    check(key, position.longitude, 'body');
  }

  if (hasHouses) {
    check('asc', chart.angles.asc, 'angle');
    check('mc', chart.angles.mc, 'angle');

    const cusps = chart.houses.get(chart.houseSystem).cusps;
    for (let house = 1; house <= 12; house += 1) {
      const orb = Math.abs(norm180(longitude - cusps[house - 1]));
      if (orb <= cuspOrb) {
        hits.push({
          transit: transitName, natal: `куспид ${house}`, natalKind: 'cusp',
          house, aspect: ASPECTS[0], orb, exact: orb <= EXACT,
          natalLongitude: cusps[house - 1],
          strength: Math.max(0, 1 - orb / cuspOrb),
        });
      }
    }
  }

  hits.sort((a, b) => a.orb - b.orb);

  const cusps = hasHouses ? chart.houses.get(chart.houseSystem).cusps : null;
  const house = cusps ? houseOf(cusps, longitude) : null;

  const rulership = hasHouses ? ruledHouses(chart) : new Map();
  const touched = new Set();
  if (house) touched.add(house);
  for (const hit of hits) {
    if (hit.natalKind === 'body') {
      for (const ruled of rulership.get(hit.natal) || []) touched.add(ruled);
      const position = chart.positions.get(hit.natal);
      if (position && hasHouses) touched.add(position.house);
    } else if (hit.natalKind === 'cusp') {
      touched.add(hit.house);
    } else if (hit.natal === 'asc') {
      touched.add(1);
    } else if (hit.natal === 'mc') {
      touched.add(10);
    }
  }

  return {
    longitude,
    hits,
    house,
    housesTouched: [...touched].sort((a, b) => a - b),
    ruledHouses: rulership,
    get touches() { return hits.length > 0 || house !== null; },
  };
}

// Загрузка календаря событий.
let calendar = null;

// Событие неба целиком. Полнолуние работает по оси: Луна с одной стороны,
// Солнце с другой, и у каждой стороны свой дом и свои касания. Поэтому для
// полнолуния градус проверяется дважды, а касание к точке карты берется
// от того светила, с которым оно сильнее (соединение с Солнцем важнее
// оппозиции к Луне, хотя это один и тот же градус). У каждого касания
// помечено светило (luminary) и градус, от которого оно идет (source).
const ASPECT_ORDER = ['conjunction', 'opposition', 'square', 'trine', 'sextile'];

export function isFullMoon(event) {
  return event?.kind === 'lunation' && event.detail?.phase === 'full';
}

export function examineEvent(chart, event) {
  const moon = examine(chart, event.longitude, event.title);
  if (!isFullMoon(event)) return moon;
  const sunLongitude = (event.longitude + 180) % 360;
  const sun = examine(chart, sunLongitude, event.title);
  const byPoint = new Map();
  const keyOf = (hit) => (hit.natalKind === 'cusp' ? `cusp ${hit.house}` : hit.natal);
  for (const hit of moon.hits) {
    byPoint.set(keyOf(hit), { ...hit, luminary: 'moon', source: event.longitude });
  }
  for (const hit of sun.hits) {
    const candidate = { ...hit, luminary: 'sun', source: sunLongitude };
    const previous = byPoint.get(keyOf(hit));
    if (!previous
      || ASPECT_ORDER.indexOf(candidate.aspect.key) < ASPECT_ORDER.indexOf(previous.aspect.key)) {
      byPoint.set(keyOf(hit), candidate);
    }
  }
  const hits = [...byPoint.values()].sort((a, b) => a.orb - b.orb);
  const housesTouched = [...new Set([...moon.housesTouched, ...sun.housesTouched])].sort((a, b) => a - b);
  return {
    longitude: event.longitude,
    sunLongitude,
    axis: true,
    hits,
    house: moon.house,
    sunHouse: sun.house,
    housesTouched,
    ruledHouses: moon.ruledHouses,
    get touches() { return hits.length > 0 || moon.house !== null; },
  };
}

export async function loadEvents(url = 'data/events.json') {
  if (calendar) return calendar;
  try {
    const response = await fetch(url);
    if (!response.ok) throw new Error(String(response.status));
    const payload = await response.json();
    calendar = payload.events || [];
  } catch (error) {
    calendar = [];
  }
  return calendar;
}

export function allEvents() {
  return calendar || [];
}

// Ближайшие события: недавно прошедшие и предстоящие.
export function eventsAround(date = new Date(), before = 21, after = 45) {
  const today = date.toISOString().slice(0, 10);
  const from = new Date(date.getTime() - before * 86400000).toISOString().slice(0, 10);
  const to = new Date(date.getTime() + after * 86400000).toISOString().slice(0, 10);
  return allEvents()
    .filter((event) => event.date >= from && event.date <= to)
    .map((event) => ({ ...event, past: event.date < today }));
}

export function findEvent(key) {
  if (!key) return null;
  // Ключ может быть с датой — так ссылка указывает на конкретное событие,
  // а не на любое такое же в календаре.
  const [base, date] = key.split('@');
  const matches = allEvents().filter((event) => event.key === base);
  if (!matches.length) return null;
  if (date) return matches.find((event) => event.date === date) || matches[0];
  // Без даты берётся ближайшее к сегодняшнему дню.
  const today = new Date().toISOString().slice(0, 10);
  return matches.reduce((best, event) => {
    const distance = Math.abs(Date.parse(event.date) - Date.parse(today));
    const bestDistance = Math.abs(Date.parse(best.date) - Date.parse(today));
    return distance < bestDistance ? event : best;
  });
}

File: astro/transits.py
"""Транзиты: как положение планеты в небе задевает натальную карту.

Транзит — это не свойство карты, а встреча: планета в небе подходит к
градусу, который в карте чем-то занят. Поэтому здесь считается не «что
происходит», а «за что именно в этой карте цепляется вот этот градус»:
какие тела он аспектирует, в какой дом попадает, какие куспиды задевает
и — через управителей — каких тем это касается.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Dict, List, Mapping, Optional, Sequence, Tuple

from . import aspects as aspects_mod
from . import rulers as rulers_mod
from .chart import Chart
from .zodiac import norm180, norm360, to_sign

#: Орбисы транзитов уже натальных: транзитная планета проходит градус за
#: дни или недели, и широкий орбис размазал бы событие на месяцы.
DEFAULT_TRANSIT_ORBS: Dict[str, float] = {
    "conjunction": 3.0,
    "opposition": 3.0,
    "square": 3.0,
    "trine": 3.0,
    "sextile": 2.0,
}

#: Орбис попадания транзита на куспид дома.
DEFAULT_CUSP_ORB = 2.0


@dataclass(frozen=True)
class TransitHit:
    """Аспект транзитной точки к точке натальной карты."""

    transit: str          # что идёт по небу
    natal: str            # к чему в карте
    natal_kind: str       # body | angle | cusp
    aspect: aspects_mod.Aspect
    orb: float
    exact: bool           # точный в пределах четверти градуса
    natal_longitude: float

    @property
    def strength(self) -> float:
        limit = DEFAULT_TRANSIT_ORBS.get(self.aspect.key, 3.0)
        return max(0.0, 1.0 - self.orb / limit) if limit else 0.0

    def describe(self) -> str:
        return (
            f"{self.transit} {self.aspect.name.lower()} "
            f"{self.natal} (орб {self.orb:.1f}°)"
        )


@dataclass(frozen=True)
class TransitReport:
    """Что транзитный градус задевает в конкретной карте."""

    longitude: float
    hits: Tuple[TransitHit, ...]
    house: Optional[int]          # в какой натальный дом попал градус
    houses_touched: Tuple[int, ...]   # дома, задетые через тела и куспиды
    ruled_houses: Mapping[str, Tuple[int, ...]]  # какое тело каким домом правит

    @property
    def touches(self) -> bool:
        return bool(self.hits) or self.house is not None

    @property
    def closest(self) -> Optional[TransitHit]:
        return self.hits[0] if self.hits else None


def _aspect_list(names: Optional[Sequence[str]]) -> Tuple[aspects_mod.Aspect, ...]:
    if names is None:
        return aspects_mod.MAJOR
    return tuple(aspects_mod.BY_KEY[name] for name in names)


def ruled_houses(chart: Chart) -> Dict[str, Tuple[int, ...]]:
    """Какими домами управляет каждое тело карты.

    Дом управляется телом, которому принадлежит знак на его куспиде. Это
    то, что переводит «задета планета» в «задета тема»: транзит к
    управителю второго дома говорит о деньгах, даже если сама планета
    стоит в седьмом.
    """
    houses = chart.primary_houses
    mapping: Dict[str, List[int]] = {}
    for house in range(1, 13):
        sign = to_sign(houses.cusp(house)).sign_index
        ruler = rulers_mod.ruler_of(sign, chart.ruler_scheme)
        mapping.setdefault(ruler, []).append(house)
    return {key: tuple(value) for key, value in mapping.items()}


def examine(
    chart: Chart,
    longitude: float,
    transit_name: str = "транзит",
    orbs: Optional[Mapping[str, float]] = None,
    aspect_names: Optional[Sequence[str]] = None,
    cusp_orb: float = DEFAULT_CUSP_ORB,
    include_angles: bool = True,
    include_cusps: bool = True,
) -> TransitReport:
    """Разбирает, за что в карте цепляется градус ``longitude``.

    Время рождения известно не всегда. Без него нет ни домов, ни углов,
    поэтому проверяются только тела: показывать дома по карте на полдень
    значило бы выдавать выдумку за расчёт.
    """
    limits = dict(DEFAULT_TRANSIT_ORBS)
    if orbs:
        limits.update(orbs)
    aspect_set = _aspect_list(aspect_names)
    has_houses = chart.moment is not None and chart.exact_time

    hits: List[TransitHit] = []

    def check(natal_name: str, natal_longitude: float, kind: str) -> None:
        for aspect in aspect_set:
            limit = limits.get(aspect.key)
            if not limit:
                continue
            delta = norm180(longitude - natal_longitude)
            target = aspect.angle if delta >= 0 else -aspect.angle
            orb = abs(norm180(delta - target))
            if orb <= limit:
                hits.append(TransitHit(
                    transit=transit_name, natal=natal_name, natal_kind=kind,
                    aspect=aspect, orb=orb, exact=orb <= 0.25,
                    natal_longitude=natal_longitude,
                ))
                break  # ближайший аспект для пары найден

    for key, position in chart.positions.items():
        check(key, position.longitude, "body")

    if include_angles and has_houses:
        check("asc", chart.angles.asc, "angle")
        check("mc", chart.angles.mc, "angle")

    if include_cusps and has_houses:
        for house in range(1, 13):
            cusp = chart.primary_houses.cusp(house)
            if abs(norm180(longitude - cusp)) <= cusp_orb:
                hits.append(TransitHit(
                    transit=transit_name, natal=f"куспид {house}",
                    natal_kind="cusp",
                    aspect=aspects_mod.CONJUNCTION,
                    orb=abs(norm180(longitude - cusp)),
                    exact=abs(norm180(longitude - cusp)) <= 0.25,
                    natal_longitude=cusp,
                ))

    hits.sort(key=lambda hit: hit.orb)

    house = chart.primary_houses.house_of(longitude) if has_houses else None

    rulership = ruled_houses(chart) if has_houses else {}
    touched: List[int] = []
    if house is not None:
        touched.append(house)
    for hit in hits:
        if hit.natal_kind == "body":
            touched.extend(rulership.get(hit.natal, ()))
            position = chart.positions.get(hit.natal)
            if position is not None:
                touched.append(position.house)
        elif hit.natal_kind == "cusp":
            touched.append(int(hit.natal.split()[-1]))
        elif hit.natal == "asc":
            touched.append(1)
        elif hit.natal == "mc":
            touched.append(10)

    return TransitReport(
        longitude=norm360(longitude),
        hits=tuple(hits),
        house=house,
        houses_touched=tuple(sorted(set(touched))),
        ruled_houses=rulership,
    )


def positions_at(eph, t, keys: Sequence[str]) -> Dict[str, float]:
    """Долготы транзитных тел на момент."""
    from . import bodies as bodies_mod
    from . import nodes as nodes_mod

    result: Dict[str, float] = {}
    for key in keys:
        body = bodies_mod.get(key)
        if nodes_mod.is_lunar_point(body):
            result[key] = nodes_mod.position(body, eph, t).longitude
        else:
            result[key] = eph.ecliptic(body, t)[0]
    return result
